usb: device_next: Simple NCM driver for usb-next

rgrr opened the pull request:
This is a first version of an NCM driver for usb-next.  As a template the usb-next ECM driver has been used.

The driver has been kept very simple (one datagram per NTB and one buffer per direction). Current idea is not to optimize performance, focus is more on existence of an NCM driver for Zephyr.

To make the driver work with iperf (my testcase), `CONFIG_NET_BUF_VARIABLE_DATA_SIZE=y` must be used.

Discussion about this PR can be found in https://github.com/zephyrproject-rtos/zephyr/issues/71451.

Resolves: #71451

josuah commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1282 @@
+	},                                                                             \
+	};                                                                             \
+	\
+	const static struct usb_desc_header *cdc_ncm_fs_desc_##n[] = {                 \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.iad,              \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0,              \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_header,       \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_union,        \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_ncm,          \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_netfun_ncm,   \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_int_ep,       \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if1_0,            \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if1_1,            \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if1_1_in_ep,      \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if1_1_out_ep,     \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.nil_desc,         \
+	};                                                                             \
```
The way descriptors are configured in other classes seems a bit different:
https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/usbd_msc.c#L86-L93
https://github.com/zephyrproject-rtos/zephyr/blob/c211cb347e0af0a4931e0e7af3d93577bcc7af8f/subsys/usb/device_next/class/usbd_cdc_acm.c#L61-L78
https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/bt_hci.c#L91-L108

But yet, at some other place in a recently added class:
https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/usbd_uac2_macros.h#L649-L651

So I am not sure...

rgrr replied:
I took the ECM driver as a template.  So anything questionable comes from there ;-)

josuah replied:
Thank you! I asked for clarification on the [chat](https://discord.gg/zephyrproject)...

josuah replied:
Ah, I figured some of it out: A main struct is defined with all the descriptors together:

https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/usbd_cdc_ecm.c#L60-L79

And then, to be able to have a variant for FullSpeed and HighSpeed, these structs fields are referenced in two different descriptors sets, an array of pointers:

https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/usbd_cdc_ecm.c#L778-L804

This makes all sense now! So this looks like the way forward to have several sets of descriptors without re-declaring everything twice, only picking different lists.

This means the way you did, and that ECM did, looks very canonical, and forward compatible with USB3 should that be implemented.

josuah replied:
This is where this new syntax got introduced:
https://github.com/zephyrproject-rtos/zephyr/commit/42f7e1b97fc7217ae972540c3574cfcedb526941

You have quite the latest code there! Already supporting bleeding edge features such as dual-speed support.

josuah commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1282 @@
+		return 0;
+	}
+
+	LOG_WRN("usbd_cdc_ncm_control_to_device - bmRequestType 0x%02x bRequest 0x%02x unsupported",
```
I think the logging subsystem for Zephyr currently handles printing the function name. It is a Kconfig.
https://docs.zephyrproject.org/latest/services/logging/index.html#global-kconfig-options
So probably no need to add it there.

rgrr replied:
Yes, one can switch this on via CONFIG_LOG_FUNC_NAME_PREFIX_*.  Only switched on for DBG by default.

You will not see this message except if the host driver does something not expected.

josuah commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1282 @@
+
+	if (data->if_state == IF_STATE_INIT) {
+		data->if_state = IF_STATE_FIRST_SKIPPED;
+		LOG_DBG("Skip first iface enable");
```
Any help for understanding this? Is there a bug elsewhere that needs a workaround, or is it handling of a special case?

rgrr replied:
Yes... and no... explanation can be found in the NCM spec, chapter 7.1.  I have to admit that this does not seem to be intuitive, but this is how NCM works

josuah replied:
> 7.1 Notification Sequencing 
> NCM functions are required to send ConnectionSpeedChange and NetworkConnection notifications in a specific order. To simplify the coding of host device drivers, functions that are going to send a NetworkConnection notification with wValue == 0001h must first send a ConnectionSpeedChange notification that  indicates the connection speed that will be in effect when the new connection takes effect. This sequencing is justified as follows.  If the ConnectionSpeedChange follows a NetworkConnection notification, then the host driver cannot signal network connection with the correct speed until the ConnectionSpeedChange is received.  This delay may introduce latency between bus events and system events, or may cause host system overhead due to a spurious change in speed.  If the function signals the connection speed first, then the host driver will know the signaling speed at the time the network connection becomes valid.

josuah replied:
It looks like it skips through the states of a state machine without looking at the content to confirm this was the right state. Is it due to this driver controlling what events are generated here?
https://github.com/zephyrproject-rtos/zephyr/pull/72310/files#diff-d6af60862ee4693cfe35b5df0806df2fdfff1921acc16a65b23ee407548e7212R546-R572

Many thanks!

rgrr replied:
The host driver sends notifications in a "special" sequence with alternate=0/1.  The sequence is (more or less) from the spec.  If the device driver is in the correct state, it has to respond first with "speed change" and then with "connected" notifications.
After that the driver is allowed to transmit bulk data.

josuah replied:
For some reason, I thought `IF_STATE_FIRST_SKIPPED` was to skip the first message: drop it without processing it, but no: it is skipping any action on the first *alternate setting*! Right?

If interested in being more dummy-proof, renaming `IF_STATE_FIRST_SKIPPED` to `IF_STATE_FIRST_ALTERNATE_SETTING_SKIPPED` or shorter that still hint about the `_FIRST_...` of what it is.

> ### 5.3 Data Interface Descriptor Requirements 
> The Data Interface of an NCM networking function shall have two alternate settings.  **The first alternate  setting** (the default interface setting, alternate setting 0) shall include no endpoints and therefore **no networking traffic can be exchanged when the default interface setting is selected**.  The second alternate setting (alternate setting 1) is used for normal operation, and shall include one bulk IN endpoint and one bulk OUT endpoint.

rgrr replied:
done (have to push)

josuah commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1282 @@
+
+
+
+static int cdc_ncm_acl_out_cb(struct usbd_class_data *const c_data,
```
This name looks much like TinyUSB, which is a great project! But is misleading in this context, as TinyUSB "callbacks" are defined at compile time, which Zephyr does not support. This suggests that this is to be passed as a function pointer, but it is not.

rgrr replied:
No, this is directly from the ECM driver.  I have used the ECM driver (same folder) as a template and just filled in the relevant NCM stuff.

josuah replied:
Oh right, I need to give a closer look at ECM. My bad!

josuah reviewed (commented):
I am nowhere close to a maintainer, and this is an unsolicited review, but I am trying to anticipate some points in hope to save you some time.

To increase the chances to have this merged sooner than later, I suggest to:
- give the coding style a look https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style eventually applying `clang-format -i` to all C source/headers to save you time manually reformatting everything, and the result will be matching the expectations.
- submit several pull-requests, for instance one with just the header definitions from the standard

Glad to see some interest in the USB stack!

Sorry for the shallow overview, I did not read CDC NCM yet...

rgrr commented on subsys/usb/device_next/class/usbd_cdc_ncm_private.h:
```diff
@@ -0,0 +1,127 @@
+/*
```
Not sure about this one: if you have a look at https://github.com/zephyrproject-rtos/zephyr/tree/main/subsys/usb/device_next/class you will notice another "private" header which was the reason for me to put the NCM internals there.  Perhaps I should rename it to "usbd_cdc_ncm_internal.h"?

josuah replied:
Yes although every USB class also seems to have a header in [include](https://github.com/zephyrproject-rtos/zephyr/tree/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/include/zephyr/usb/class)... Except CDC ECM!

rgrr replied:
But those header contain some kind of API.  The NCM structs are internal and of no interest to externals.  So they should be private/internal.

josuah replied:
Ah understood! It looks like there is only a special case for CDC only, so maybe it is good to keep the declarations as local as possible...
https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/include/zephyr/usb/class/usb_cdc.h#L149-L211

I am curious to read the result of it.

rgrr replied:
hmmm... not sure if I got your point...

Most of the descriptor stuff seems to be in usb_ch9.h.  The class headers are one level below of that and contain class dependent stuff of the descriptors plus some API.  So usb_cdc.h with its descriptor definitions is at the right place.

But nevertheless this is not completely clear, because e.g. msos_desc.h is at the same level as usb_ch9.h but clearly belongs to MSC.  So perhaps everything is experimental and needs some cleanup.

tmon-nordic replied:
> But nevertheless this is not completely clear, because e.g. msos_desc.h is at the same level as usb_ch9.h but clearly belongs to MSC. So perhaps everything is experimental and needs some cleanup.

I find it really surprising that you mention that `msos_desc.h` "clearly belong to MSC", because it does not. It has really nothing to do with MSC and MSC device wouldn't benefit at all from `msos_desc.h`. The real benefit for `msos_desc.h` is to allow device to tell Windows that it wants WinUSB driver automatically installed. This is not needed at all for MSC, because MSC uses well-known Mass Storage class driver.

rgrr replied:
Sorry, yes you are right of course.  Mixed several things up and to make it worse I have used in one of my projects this WinUSB mechanism to have "driverless" setup (i.e. without fiddling) under Win10.

Nevertheless the file should be still in the class subdirectory, or not?

rgrr commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1282 @@
+
+
+
+static void _usbd_cdc_ncm_notification_next_step(struct usbd_class_data *const c_data)
```
renamed the three functions to z_*.  Idea behind naming is actually not the static keyword because almost everything is static in the driver.  It is to distinguish between API relevant functions and internal things

josuah replied:
I am also a bit puzzled about what to do I admit... more context: https://github.com/zephyrproject-rtos/zephyr/issues/58007

rgrr replied:
Haha, so this recommends to revert to "_" as a prefix.  I will do this on request or on the next required change.

rgrr replied:
ok... reverted back to "_"

rgrr commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1282 @@
+				sizeof(ncm_notify_connected));
+		LOG_DBG("cdc_ncm_send_notification_connected %d", ret);
+	}
+}   /* _usbd_cdc_ncm_notification_next_step */
```
Not my intention :-)  it is just my personal way and based on my "browsing through source code" experience to make things easier.

rgrr commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1282 @@
+	struct cdc_ncm_eth_data *const data = dev->data;
+	struct usbd_cdc_ncm_desc *desc = data->desc;
+
+	LOG_DBG("");
```
Yes of course.  The driver currently lacks some relevant features and it is in experimental state.  So I decided to leave those traces in, except anyone wishes otherwise.

rgrr commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1282 @@
+
+static int cdc_ncm_acl_out_cb(struct usbd_class_data *const c_data,
+		struct net_buf *const buf, const int err)
+/**
+ * Frame received from host!
+ */
```
I did not find this in style guides.  In eclipse doing it this way is more convenient because the pop-up hover of a function then first shows function name incl parameters and then the explanation.  Handy at least for me.

If you convince me otherwise I will change it.

josuah replied:
I was looking at other Zephyr source file.

But it looks like the Linux kernel coding style does not say anything about it! https://kernel.org/doc/html/latest/process/coding-style.html

I might have mistaken with http://man.openbsd.org/style.9 :)

Not up to me to tell anyway!

rgrr replied:
ok, so let's wait and see.  Compliance check is also no longer complaining

rgrr commented:
> I am nowhere close to a maintainer, and this is an unsolicited review, but I am trying to anticipate some points in hope to save you some time.

Many thanks for reviewing the code!

>     * give the coding style a look https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style eventually applying `clang-format -i` to all C source/headers to save you time manually reformatting everything, and the result will be matching the expectations.

Currently working on this.  Iterating thru the code until the runner no longer complains (which seems to be a painful process)

>     * submit several pull-requests, for instance one with just the header definitions from the standard

Unfortunately all those changes depend on each other.  So there is no meaning of providing e.g. just the headers.

josuah commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1282 @@
+ * Copyright (c) 2017 Intel Corporation
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ * Copyright (c) 2024 Hardy Griech
+ *
```
I see you contributed it altogether in TinyUSB too, so this solves the problem. Thank you for clarifying!

josuah commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1278 @@
+#include <zephyr/net/net_pkt.h>
+#include <zephyr/net/ethernet.h>
+
+#include <eth.h>
```
Is this `<zephyr/drivers/ethernet/eth.h>`? Was there a reason for leaving-off the prefix? Just curious.

rgrr replied:
not sure, but again... taken directly from the ECM driver :-)

josuah commented:
I realize that a lot of my comments might as well be addressed to USB CDC ECM, and open questions for the USB stack in general. In hope you can get more folks offering some review soon. I can ping the [`#pr-help`](https://discord.gg/zephyrproject) channel to draw attention on a pull request that needs some review when you wish.

josuah commented:
I do like how this class is a thin wrapper between the Ethernet and USB Class APIs.

rgrr commented:
> I realize that a lot of my comments might as well be addressed to USB CDC ECM, and open questions for the USB stack in general. In hope you can get more folks offering some review soon. I can ping the [`#pr-help`](https://discord.gg/zephyrproject) channel to draw attention on a pull request that needs some review when you wish.

Yes please, that would be very kind.

rgrr commented:
> I do like how this class is a thin wrapper between the Ethernet and USB Class APIs.

The ECM driver is even more direct (BTW Johann Fischer is the author of that).  NCM has the capability to additionally packing multiple datagrams into one transfer block.  But as stated in the intro this is currently not implemented.

Actually for me the nice thing is, that there are NCM drivers even for Win10.

josuah commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1278 @@
+
+	usbd_ep_enqueue(c_data, buf);
+	k_sem_take(&data->sync_sem, K_FOREVER);
+	net_buf_unref(buf);
```
If I understand correctly, this is *unref* and not *free*, so the `usbd_ep_enqueue()` still holds a reference to this buffer, and is likely not freed at this moment.  The other `unref()` are after error management.

The Ethernet API takes care of free-ing (unref-ing) the buffer itself when done.

Does it match how you sees it too?

rgrr replied:
Again taken from ECM ;-)  And it is working too.  Would be very fast become obvious if there is a buffer leak.

josuah replied:
No doubt! And it also makes sense on the source: the imported part looks all great.

rgrr replied:
yes, it does... almost just inserted on TX side framing and on RX side verification... plus a lot of renaming... plus the extra logic in the update handler... plus the request handler... ;-)

josuah commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1278 @@
+	}
+
+	usbd_ep_enqueue(c_data, buf);
+	k_sem_take(&data->sync_sem, K_FOREVER);
```
Could you hint me why this is required? It seems released here: https://github.com/zephyrproject-rtos/zephyr/pull/72310/files#diff-d6af60862ee4693cfe35b5df0806df2fdfff1921acc16a65b23ee407548e7212R598

If it is for managing the concurrency that `usbd_ep_enqueue()` might generate, should it be placed one line before to handle a context switch between `usbd_ep_enqueue()` and `k_sem_take()`?

I think I simply misunderstood something.

rgrr replied:
Again: taken directly from ECM.  It blocks the thread in line 930 until the semaphore is given in the completion event handler usbd_cdc_ncm_request().

As said, the driver is currently very simple and not performance trimmed.

josuah replied:
Thank you! This suggests it was already reviewed/validated earlier.

josuah commented:
You might be asked to merge several commits together. If that were to happen, do feel free to reach me at me@josuah.net  (or `josuah.demangeon` on the [chat](https://discord.gg/zephyrproject)) so that we can get this as expected by the real reviewers.

jukkar commented:
Please fix the commit subjects, now they are all the same which is not good. The commit subject should indicate what the commit is changing.

rgrr commented:
> Please fix the commit subjects, now they are all the same which is not good. The commit subject should indicate what the commit is changing.

Done that and squashed everything.  Top be honest, commit message policy is not clear to me.  If one does not use the special subject structure, then one of the CI checks complains, also it seems that only squashed PRs are merged, so the history of a PR seems to be irrelevant.

jukkar commented:
> it seems that only squashed PRs are merged, so the history of a PR seems to be irrelevant.

This is definitely not true, the commit history is very much relevant. The commit messages rules are actually very simple, the individual commits should contain logical changes related to each other. For new code, in many cases there is need to have only couple of commits, for example the actual feature in one, tests in another, samples in another. Sometimes it makes sense to split the feature to multiple commits each implementing sub-features. It really depends what is being submitted.

rgrr commented:
> > it seems that only squashed PRs are merged, so the history of a PR seems to be irrelevant.
> 
> This is definitely not true, the commit history is very much relevant. The commit messages rules are actually very simple, the individual commits should contain logical changes related to each other. For new code, in many cases there is need to have only couple of commits, for example the actual feature in one, tests in another, samples in another. Sometimes it makes sense to split the feature to multiple commits each implementing sub-features. It really depends what is being submitted.

Thanks for the clarification, did not want to offend anybody.

I have to admit that I'm new to Zephyrs PR policy and thus have to get used to it.

josuah commented:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines

josuah commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1278 @@
+	uint8_t ep;
+	int ret;
+
+	LOG_DBG("");
```
I suppose this would receive similar comment to this:
https://github.com/zephyrproject-rtos/zephyr/pull/53408#discussion_r1182371092

A `git commit --amend` adding some short description of just a few words would allow keeping these.

rgrr replied:
removed every empty LOG_DBG(), wondering if anybody ever made the cost/environmental calculation of these CI runs caused by mini changes...

josuah replied:
I read somewhere that scripts/rules were setup to avoid extra CI cycles, with manual or auto-run CI.

Maybe there can be rules to only auto-run CI when the PR is not in "draft" state.

josuah reviewed (commented):
This is a recent modification on upstream Zephyr: `usbd_contex` had a typo and was adjusted to `usbd_context`.

https://github.com/zephyrproject-rtos/zephyr/commit/8d344cc9d8484767f4d72a0835b5e12a614b5e58

josuah commented:
To help with maintenance and troubleshooting, I believe Zephyr maintainers like to have only meaningful commits.
Maybe making a single commit that introduces the class would work?

`git rebase -i HEAD^4` and modifying the lines for the two commits that needs to be changed from `pick` to `squash` would likely work.

rgrr commented:
> This is a recent modification on upstream Zephyr: `usbd_contex` had a typo and was adjusted to `usbd_context`.
> 

Haha... first time I couldn't compile before pushing, because my environment was broken (see https://github.com/eclipse-cdt/cdt/pull/846).  And then the CI shows you that that's a bad idea.

Squashed the commits (again).  I actually don't like that, because history is lost (and actually I'm not a fan of rebasing because history is rewritten, but that is another story...))

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1270 @@
+	const struct nth16_t *nth16 = &(ntb->nth);
+	const uint16_t len = buf->len;
+
+	LOG_DBG("%p, %d", ntb, (int)len);
```
LOG_DBG("Fancy text %p, %u", ntb, len);

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1270 @@
+	 * check header
+	 */
+	if (len < sizeof(ntb->nth)) {
+		LOG_ERR("  ill length: %d", len);
```
Why are there spaces before ill? Please remove it does not make any sense.
Is ill, ill or illegal?

rgrr replied:
It is "illegal"... I will correct it accordingly

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1270 @@
+	if (len < sizeof(ntb->nth)) {
+		LOG_ERR("  ill length: %d", len);
+		return false;
+	}
+	if (sys_le16_to_cpu(nth16->wHeaderLength) != sizeof(struct nth16_t)) {
```
```c 
	}

	if (sys_le16_to_cpu(nth16->wHeaderLength) != sizeof(struct nth16_t)) {
```
Please apply everywhere.

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1270 @@
+	}
+	if (sys_le32_to_cpu(nth16->dwSignature) != NTH16_SIGNATURE) {
+		LOG_ERR("  ill signature: 0x%08x",
+				(unsigned int)sys_le32_to_cpu(nth16->dwSignature));
```
Why is there explicit type conversion?

rgrr replied:
habit of mine to avoid messages on different compilers because fmt-string does not fit arguments

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1270 @@
+				(unsigned int)sys_le32_to_cpu(nth16->dwSignature));
+		return false;
+	}
+	if (len < sizeof(struct nth16_t) + sizeof(struct ndp16_t)
+			+ 2*sizeof(struct ndp16_datagram_t)) {
```
What is `sizeof(struct nth16_t) + sizeof(struct ndp16_t)	+ 2*sizeof(struct ndp16_datagram_t)` ?

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1270 @@
+		return false;
+	}
+	if (sys_le16_to_cpu(nth16->wBlockLength) > CFG_CDC_NCM_RCV_NTB_MAX_SIZE) {
+		LOG_ERR("  ill block length2: %d > %d", sys_le16_to_cpu(nth16->wBlockLength),
+				CFG_CDC_NCM_RCV_NTB_MAX_SIZE);
```
```c

		LOG_ERR("Wrong block length2: %u > %u",
			sys_le16_to_cpu(nth16->wBlockLength), CFG_CDC_NCM_RCV_NTB_MAX_SIZE);
```

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1270 @@
+	/*
+	 * check (first) NDP(16)
+	 */
+	const struct ndp16_t *ndp16 = (const struct ndp16_t *)
```
Variable definition should be at the top of a block. Please clean up and fix your code, I will continue the review after that.

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1270 @@
+
+
+
+static bool _cdc_ncm_frame_ok(struct cdc_ncm_eth_data *data, struct net_buf *const buf)
```
Prefixing with **_** is not okay.

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm_internal.h:
```diff
@@ -0,0 +1,128 @@
+#define CFG_CDC_NCM_RCV_NTB_MAX_SIZE              3200
+
+/* Table 6.2 Class-Specific Request Codes for Network Control Model subclass */
+typedef enum {
```
No typedefs.

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm_internal.h:
```diff
@@ -0,0 +1,128 @@
+#define __USBD_CDC_NCM_LOCAL_H_
+
+
+#ifndef CFG_CDC_NCM_ALIGNMENT
+#define CFG_CDC_NCM_ALIGNMENT                 4
+#endif
+#if (CFG_CDC_NCM_ALIGNMENT != 4)
+/*  headers and start of datagrams in structs have to be aligned differently */
+#error "CFG_CDC_NCM_ALIGNMENT must be 4"
+#endif
```
Is it not the same as just `#define CFG_CDC_NCM_ALIGNMENT                 4`?
Please remove all the `CFG_` prefixes and move whatever make sense to Kconfig options.

rgrr replied:
I have "reserved" those CFGs for future extensions.  Currently there is no meaning in them, so they don't appear in Kconfig.  Will change it.

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_ncm_internal.h:
```diff
@@ -0,0 +1,128 @@
+
+#define NTH16_SIGNATURE      0x484D434E
+#define NDP16_SIGNATURE_NCM0 0x304D434E
+#define NDP16_SIGNATURE_NCM1 0x314D434E
```
```c
#define NTH16_SIGNATURE      0x484D434EUL
#define NDP16_SIGNATURE_NCM0 0x304D434EUL
#define NDP16_SIGNATURE_NCM1 0x314D434EUL
```

rgrr replied:
@jfischer-no : many thanks for the extensive review, I will change accordingly (or comment on it).  But this will take some time.

rgrr commented on doc/connectivity/usb/device_next/usb_device.rst:
```diff
@@ -41,6 +41,8 @@ instance (`n`) and is used as an argument to the :c:func:`usbd_register_class`.
 +-----------------------------------+-------------------------+-------------------------+
 | USB CDC ECM class                 | Ethernet device         | :samp:`cdc_ecm_{n}`     |
 +-----------------------------------+-------------------------+-------------------------+
+| USB CDC NCM class                 | Ethernet device         | :samp:`cdc_ncm_{n}`     |
```
ok... I have created three branches and soon two more PRs will appear here.  New code stays in this PR.

Concerning review procedure: who is resolving conversations?  You?  Me?

tmon-nordic replied:
One branch (one PR) with N commits is the preferred way.

rgrr replied:
gosh... what have I done!?  Misunderstood Johanns point completely.  Will go back.

rgrr replied:
what commit comments do you expect in this case?

tmon-nordic replied:
If by commit comments you mean the commit messages, then each commit message should describe what the individual commit does. Do not reuse the title for all of the commits, make each of them point to the respective part. For example the commit changing sample should start its commit title with `samples: net: zperf:`.

Things like "This PR contains the actual driver." do not belong in commit message.

rgrr commented on samples/net/zperf/overlay-netusb.conf:
```diff
@@ -1,6 +1,12 @@
 # USB Device Settings
 CONFIG_USB_DEVICE_STACK=y
 
+CONFIG_NET_PKT_RX_COUNT=14
+CONFIG_NET_PKT_TX_COUNT=14
+CONFIG_NET_BUF_RX_COUNT=28
+CONFIG_NET_BUF_TX_COUNT=28
+CONFIG_NET_BUF_DATA_SIZE=1500
+
```
I have to rethink why I have done this ;-)

rgrr commented on samples/net/zperf/overlay-usbd_next.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_USB_DEVICE_STACK_NEXT=y
+
+# next-ncm does not work well with the static-size buffers
+CONFIG_NET_PKT_BUF_RX_DATA_POOL_SIZE=10000
+CONFIG_NET_PKT_BUF_TX_DATA_POOL_SIZE=10000
+CONFIG_NET_BUF_VARIABLE_DATA_SIZE=y
+
+CONFIG_LOG=y
+CONFIG_USBD_LOG_LEVEL_WRN=y
+CONFIG_UDC_DRIVER_LOG_LEVEL_WRN=y
```
done that

rgrr commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -0,0 +1,1270 @@
+	return desc->if0.bInterfaceNumber;
+}   /* cdc_ncm_get_ctrl_if */
+
+
+
```
The code is heavily based on the ECM driver.  Sorry for having my own "style"

rgrr commented on include/zephyr/usb/class/usb_cdc.h:
```diff
@@ -199,7 +202,7 @@ struct cdc_acm_notification {
 
 /** Ethernet Networking Functional Descriptor */
-struct cdc_ecm_descriptor {
+struct cdc_eth_functional_descriptor {
```
Hmm... but `cdc_ecm_descriptor` is the wrong name.  Will change it back anyway

rgrr committed:
usb: device_next: Simple NCM driver for usb-next

Documentation changes.

Signed-off-by: Hardy Griech <ntbox@gmx.net>

rgrr committed:
usb: device_next: Simple NCM driver for usb-next

Sample changes.

Signed-off-by: Hardy Griech <ntbox@gmx.net>

rgrr committed:
usb: device_next: Simple NCM driver for usb-next

This is a simple USB-NCM driver. It just has one NTB
per direction and within the NTB only one datagram
can be received/transmitted.
Main goal is existance of an NCM driver, performance
will follow in further steps.

This PR contains the actual driver.

Signed-off-by: Hardy Griech <ntbox@gmx.net>

jfischer-no commented:
@rgrr There still seems to be no changes since the last review, do you have any challenges with git?

rgrr commented:
> @rgrr There still seems to be no changes since the last review, do you have any challenges with git?

No... sorry... I'm currently too busy to work on the PR.  Don't drop it, if there is time I will polish it.

josuah commented:
Following this initial effort on bringing CDC NCM to Zephyr, an independent PR was started after this message and is now merged:
- https://github.com/zephyrproject-rtos/zephyr/pull/79508

The outline of functions is the same, but their implementation differ in several places.
I suppose this PR content helped in giving insight in how the problem could be solved, and was supportive on the other PR which led to CDC NCM merged Zephyr.

Thank you @rgrr !
It might be possible to close this PR now, as the NCM development effort were pursued on the other PR.

rgrr commented:
:
> The outline of functions is the same, but their implementation differ in several places. I suppose this PR content helped in giving insight in how the problem could be solved, and was supportive on the other PR which led to CDC NCM merged Zephyr.

Great, if my implementation was of any help!

> Thank you @rgrr ! It might be possible to close this PR now, as the NCM development effort were pursued on the other PR.

Of course close it.  I will check the other implementation soon and provide comments if required.

Sorry for wasting your time for reviewing my PR.

josuah commented:
> Sorry for wasting your time for reviewing my PR
No issue at all ! I learned in the process and that was useful in other ways (UVC).
